### PR TITLE
DRIVERS-2573: Reference test for $$matchAsDocument and $$matchAsRoot

### PR DESCRIPTION
**Pull request by jmikola**
https://jira.mongodb.org/browse/DRIVERS-2573

Please complete the following before merging:

- [x] Test changes in at least one language driver: https://github.com/mongodb/mongo-php-library/pull/1508

Note: I implemented `$$matchAsDocument` and `$$matchAsRoot` in PHPLIB in order to POC this, so I'd appreciate it if another language driver could also test this.

### CODE REVIEW

**Review by alcaeus — APPROVED** [review]

File: source/unified-test-format/tests/valid-fail/operator-matchAsDocument.yml [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+      id: &collection0 collection0
+      database: *database0
+      collectionName: &collection0Name coll0
+
+initialData:
+  - collectionName: *collection0Name
+    databaseName: *database0Name
+    documents:
+      - { _id: 1, json: '{ "x": 1, "y": 2 }' }
+      - { _id: 2, json: '{ "x" }' }
```
Comment (GromNaN) [inline source/unified-test-format/tests/valid-fail/operator-matchAsDocument.yml:22]: You can add an error case where `json` is a json array.
```suggestion
      - { _id: 2, json: '{ "x" }' }
      - { _id: 3, json: '[{"x":1}]' }
```

It seems to be in your implementation: https://github.com/mongodb/mongo-php-library/pull/1508/files#diff-5911ac585872dc4a55ff2fe77af552d46e160b4d5636699c9cf2c1211217e25eR278

**Review by ShaneHarvey — COMMENTED** [review]
Tests pass locally in Python, running a patch here: https://github.com/mongodb/mongo-python-driver/pull/1988

This uncovered a bunch of issues with Python's implementation of $$matchAsDocument and $$matchAsRoot. Nice work!

**Review by ShaneHarvey — APPROVED** [review]
LGTM!

File: source/unified-test-format/tests/valid-fail/operator-matchAsDocument.yml (hunk 1 above)
Comment (jmikola) [reply source/unified-test-format/tests/valid-fail/operator-matchAsDocument.yml]: Added some additional valid-fail tests with unexpected JSON strings.

**Comment by jmikola** [comment]
@ShaneHarvey: I made some test revisions based on @GromNaN's feedback. Also removed a useless test for `$$matchAsRoot`. Please queue up another CI run and let me know if all is well.

**Review by ShaneHarvey — APPROVED** [review]

**Pull request merged**